keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+	Spec core.ResourceClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ResourceClaimStatus
+}
+
+// ResourceClaimSpec defines how a resource is to be allocated.
+type ResourceClaimSpec struct {
+	// ResourceClassName references the driver and additional parameters
+	// via the name of a ResourceClass that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists, but a
+	// driver-specific admission webhook may require that and is allowed to
+	// reject claims where the class is missing.
+	ResourceClassName string
+
+	// Parameters references a separate object with arbitrary parameters
+	// that will be used by the
+	// driver when allocating a resource for the claim.
+	//
+	// The object must be in the same namespace as the ResourceClaim.
+	// An empty namespace field defaults to that namespace.
+	// ApiVersion and name must be set. Kind and/or resource must be set.
+	// Usually, kind is sufficient because there will be a 1:1 mapping
+	// to resource. If UID is set, then only that specific object
+	// may be used.
+	Parameters v1.ObjectReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. Waiting for a Pod is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod (AllocationModeDelayed). Other modes
+// might get added in the future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeDelayed, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeDelayed AllocationMode = "Delayed"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started. It's necessary to support
+	// deallocation when the class gets deleted before a claim.
+	DriverName string
+
+	// Scheduling contains information that is only relevant while the
+	// scheduler and the resource driver are in the process of selecting a
+	// node for a Pod and the allocation mode is AllocationModeDelayed. The
+	// resource driver should unset this when it has successfully allocated
+	// the resource.
+	Scheduling SchedulingStatus
```

Comment:
This could work, I am just not sure whether it is better. Scheduler and resource driver controller would still need write access to ResourceClaimStatus, so it's not more secure.

Performance overall may or may not be better: other clients will not see ResourceClaim status updates that they are not interested in (good), but it also creates more load for managing these extra objects (bad).

Reply:
I'm not at all convinced.  What's going on here is a complex (andf this is v0!) negotiation, storing all of its intermediate state in a user-visible resource.  It's UGLY at best.  Given how many writes will happen anyway, creating an object to host it doesn't seem to me to be a real risk, but the complexity of this API does.

Pretend you are a user who just wants to use a GPU, and you find *all this* in the docs for the API.

Reply:
Yes, let's put this into a separate object.